Workiva/wf-uicomponents
# HY-1998 Emit Event When Scrolling Beyond Viewport

JIRA >> https://jira.atl.workiva.net/browse/HY-1998
@timmccall-wf @tomconnell-wf Code Review
Testing depends on [wf-common/HY-2651_Add_Scroll_Detection](https://github.com/Workiva/wf-common/pull/42)

## Problem Summary
**Scenario: Flow Mode**

Given the top of a document is in view
When I attempt to scroll beyond the top of the document
Then an event is emitted

Given the bottom of a document is in view
When I attempt to scroll beyond the bottom of the document
Then an event is emitted

**Scenario: Peek/Single Modes**

Given the first page of a document is in view
And the top of the page is visible
When I attempt to scroll to the previous page
Then an event is emitted

Given the last page of a document is in view
And the bottom of the page is in view
When I attempt to scroll to the next page
Then an event is emitted

**Notes**
- "scroll" means both mouse wheel scrolling and swiping


## Solution
Added logic to BoundaryInterceptor which emits an event when it detects a swipe or scroll above a certain threshold beyond a boundary of the AwesomeMap.  This event is picked up and propagated by the ScrollList, and then by DocViewer.

## Unit tests
Tests added for BoundaryInterceptor and ScrollList

## How to QA/+10
1. Bower link wf-common into wf-uicomponents and check out the [related branch](https://github.com/Workiva/wf-common/pull/42)
2. Check out this branch and run `grunt qa` or`grunt serve`
3. Open the ScrollList component example
4. (In chrome) Open Dev tools -> Source and throw a breakpoint at `src/scroll_list/AwesomeMapFactory.js:171`, the createListMap function and then refresh to catch the breakpoint.
5. In the console, create a callback function to subscribe to the new onScrollPastBoundary event on ScrollList in order to monitor the events dispatched.  For example, I used: 
`
scrollList.onScrollPastBoundary(function(source, args) {
    console.log("EVENT TRIGGERED!  "+args.boundary);
});
`
6. Scroll, drag and swipe around in the window!  Try different modes, orientations, etc.  Dont forget device modes too!  Try out the touch handlers.  Make sure you can see all 4 events (left,right,up,down).
**Note:** The events have a configurable sensitivity threshold for firing, so that light touches and small scrolls (by default) wont fire events constantly.  So if you have trouble with an event not firing, try scrolling harder!  But let me know if anything is unreasonably troublesome.
